redis/shared.go

```diff
@@ -0,0 +1,76 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/bshramin/goxy"
+	"github.com/go-redis/redis/v8"
+	"github.com/go-redsync/redsync/v4"
+	"github.com/go-redsync/redsync/v4/redis/goredis/v8"
+	"github.com/sirupsen/logrus"
+)
+
+var keysList = make(map[*redis.Client]map[string]*redsync.Mutex)
+
+// this is used when multi instances of a code try to use shared cache on redis
+// and result fetch function has too much load expensive
+// it will return the same result for all of them
+// it will checks if the value exists it will return it
+// if not it will check if any one else is trying to fetch value
+// if it was the case it will not try to fetch
+// but if no one of instances has tried to fetch the result it start to fetch and set a waitKey
+// too inform others
+func SharedFetch[K any](ctx context.Context, client *redis.Client, key string, t time.Duration, f func() (K, error)) (K, error) {
+	var res K
+	redisRes := client.Get(ctx, key)
+	go fetch(ctx, client, key, t, f)
+	err := redisRes.Err()
+	if err != nil {
+		return res, err
```

Comment:
If we get an error, we would probably want to wait for the fetching to complete and then try again.